### PR TITLE
[#10455] Add a redirect banner for users to switch versions

### LINKED ISSUE #10455: Add option for users to revert to v6
Once v7 is deployed, users will use v7 as a default mode. We should add an option to revert to v6 for instructors who prefer the old user interface.

**Proposed implementation**: 
Add a banner at the top of the page (under the nav bar) with a button that users can click to be redirected to the URL of v6
- The banner will only be added to the main home page and the instructor home page
- The button will redirect to the homepage of a concurrently hosted (i.e. same database) deployment of v6
- Redirection URL configuration will be in `config.ts`

Alternative implementations considered:
- Add a button to the navbar (i.e. no banner) and an informational modal
- Have the banner persist to all instructor pages
- Redirection in the footer / banner at the bottom

Justification for proposed implementation:
- Having a persistent banner/button will likely be annoying. Even if it were dismissible, it would only be dismissed per user session. Implementing cross session tracking is going to be more work than believed necessary for this.
- Users might believe the button will take them to corresponding page in v6 which will require page matching. I'm not sure how feasible this is and believe it to be more work than necessary.
- Having a banner allows for a short message to be included (i.e. "You are viewing the new version of TEAMMATES. To return to the old version click here").
- Having a button embedded in the navbar would require that the navbar track which page it is on as it is its own component. I figured a banner would be a simpler and more straightforward implementation

### PR DESCRIPTION
**Pull request by madanalogy**
Fixes #10455 

Screenshot 2020-07-28 at 3 25 17 PM

### CODE REVIEW

**Comment by damithc** [comment]
Can give a screenshot?
How to set the target URL for the old version? Currently `6-16-4-dot-teammatesv4. ...`

**Comment by madanalogy** [comment]
> Can give a screenshot?

Screenshots added to PR description.

> How to set the target URL for the old version? Currently `6-16-4-dot-teammatesv4. ...`

After this PR is merged, you can set it in the `config.ts` file. This file is not tracked by git, and the template for it is available in `src/web/environments`

**Comment by damithc** [comment]
> After this PR is merged, you can set it in the `config.ts` file. This file is not tracked by git, and the template for it is available in [`src/web/environments`](https://github.com/TEAMMATES/teammates/blob/master/src/web/environments/config.template.ts)

Sound good.

For the message, perhaps something like this?

> You are using the revamped user interface (beta) of TEAMMATES, released mid-2020. If you would like to go back to the previous user interface, ...

**Comment by madanalogy** [comment]
> For the message, perhaps something like this?
> 
> > You are using the revamped user interface (beta) of TEAMMATES, released mid-2020. If you would like to go back to the previous user interface, ...

The main description screenshot has been edited to reflect the new message

**Comment by damithc** [comment]
> The main description screenshot has been edited to reflect the new message

👍

**Comment by damithc** [comment]
> You need to mention:
> 
> * The old one will be gone anyway, probably in about 1-2 months. Stopping the message there implies that we're going to maintain both versions, which obviously we will not.
> * (especially important) We welcome feedback on the new user interface.

How about this?

>This is the **revamped user interface (beta)** of TEAMMATES (released mid-2020) -- faster and more mobile-friendly than the previous version. **We welcome [your feedback]() on the new interface**. You can keep using the previous user interface (to be phased out by Dec 2020) by going [here]().

**Comment by wkurniawan07** [comment]
> to be phased out by Dec 2020

I'd say to be phased out by Oct 2020. Not phasing out V6 blocks us from any DB-modifying new features.

**Comment by madanalogy** [comment]
>**We welcome your feedback on the new interface**

@damithc where do you want the feedback link configured to redirect to? Or would you like to configure it in the same way as the old version redirect (i.e. in `config.ts`)?

**Comment by damithc** [comment]
> @damithc where do you want the feedback link configured to redirect to? Or would you like to configure it in the same way as the old version redirect (i.e. in `config.ts`)?

It should be same as _crash report email_ in build.properties i.e., `teammates@comp.nus.edu.sg`
Can put in config.ts if cannot get it from the backend

File: src/web/app/components/redirect-banner/redirect-banner.component.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+export class RedirectBannerComponent {
+
+  redirectUrl: SafeResourceUrl | null;
+  supportEmail: SafeResourceUrl | null;
+
+  constructor(private sanitizer: DomSanitizer) {
+    this.redirectUrl = environment.redirectUrl
+        ? this.sanitizer.bypassSecurityTrustResourceUrl(environment.redirectUrl)
+        : null;
+    this.supportEmail = environment.supportEmail
```
Comment (wkurniawan07) [inline src/web/app/components/redirect-banner/redirect-banner.component.ts:22]: Don't need this. It is reasonable to assume that `environment.supportEmail` will always be set.

File: src/web/environments/config.template.ts [hunk 2]
```diff
@@ -23,4 +23,9 @@ export const config: any = {
    * e.g. "6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI" is a site key for test environments.
    */
   captchaSiteKey: '',
+
+  /**
+   * This is the URL address that will show in the {@link RedirectBannerComponent}.
```
Comment (wkurniawan07) [inline src/web/environments/config.template.ts:28]: A config point should not be dependent of this kind of specific component name.

File: src/web/app/components/redirect-banner/redirect-banner.component.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  styleUrls: ['./redirect-banner.component.scss'],
+})
+export class RedirectBannerComponent {
+
+  redirectUrl: SafeResourceUrl | null;
+  supportEmail: SafeResourceUrl | null;
+
+  constructor(private sanitizer: DomSanitizer) {
+    this.redirectUrl = environment.redirectUrl
+        ? this.sanitizer.bypassSecurityTrustResourceUrl(environment.redirectUrl)
```
[the remark this answers is not in the document]
Comment (madanalogy) [reply src/web/app/components/redirect-banner/redirect-banner.component.ts:20]: I'm just following the example of the `accountRequestFormUrl`

https://github.com/TEAMMATES/teammates/blob/e2eb385e1d6fc5e6a18303c30bb456c8da6beee0/src/web/app/pages-static/request-page/request-page.component.ts#L18-L20

[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/web/app/components/redirect-banner/redirect-banner.component.ts:20]: That URL is used to load an iframe, so it's a legitimate security concern if not bypassed. It's not the case here.

**Comment by damithc** [comment]
@madanalogy I deployed this version but I don't see the banner. Anything I need to do first?

**Comment by madanalogy** [comment]
> @madanalogy I deployed this version but I don't see the banner. Anything I need to do first?

Yes you need to add the `redirectUrl` attribute in `config.ts`. You can refer to the [`config.template.ts`](https://github.com/TEAMMATES/teammates/blob/master/src/web/environments/config.template.ts) on how your `config.ts` file should look like

**Comment by damithc** [comment]
> Yes you need to add the `redirectUrl` attribute in `config.ts`. You can refer to the [`config.template.ts`](https://github.com/TEAMMATES/teammates/blob/master/src/web/environments/config.template.ts) on how your `config.ts` file should look like

Thanks. It's working now.

**Pull request merged**